src/phase-management/PhaseAdvancer.js

```diff
@@ -0,0 +1,249 @@
+
+    if (delta !== 0) {
+      this.#updateSubsequentPhases(phases, phase, -delta);
+    }
+
+    console.log(`Updated phases: ${JSON.stringify(phases, null, 2)}`);
+  }
+
+  async #close(challengeId, phases, phase) {
+    console.log(`Closing phase ${phase.name} for challenge ${challengeId}`);
+
+    phase.isOpen = false;
+    const actualEndDate = new Date();
+    phase.actualEndDate = actualEndDate.toISOString();
+
+    const scheduledEndDate = parseDate(phase.scheduledEndDate);
+    const delta = scheduledEndDate - actualEndDate;
+
+    if (delta !== 0) {
+      this.#updateSubsequentPhases(phases, phase, -delta);
+    }
+
+    console.log(`Updated phases: ${JSON.stringify(phases, null, 2)}`);
+  }
+
+  #isPastTime(dateString) {
+    if (dateString == null) {
+      return false;
+    }
+    const date = new Date(dateString);
+    const now = new Date();
+    return date <= now;
+  }
+
+  async #getRegistrantCount(challengeId) {
+    console.log(`Getting registrant count for challenge ${challengeId}`);
+    // TODO: getChallengeResources loops through all pages, which is not necessary here, we can just get the first page and total count from header[X-Total]
+    const submitters = await helper.getChallengeResources(challengeId, config.SUBMITTER_ROLE_ID);
+    return submitters.length;
+  }
+
+  async #getSubmissionCount(challengeId) {
+    console.log(`Getting submission count for challenge ${challengeId}`);
+    // TODO: getChallengeSubmissions loops through all pages, which is not necessary here, we can just get the first page and total count from header[X-Total]
+    const submissions = await helper.getChallengeSubmissions(challengeId);
+    return submissions.length;
+  }
+
+  async #areAllSubmissionsReviewed(challengeId) {
+    console.log(`Getting review count for challenge ${challengeId}`);
+    return Promise.resolve(false);
+  }
+
+  async #areAllAppealsResolved(challengeId) {
+    console.log(`Checking if all appeals are resolved for challenge ${challengeId}`);
+    return Promise.resolve(false);
+  }
+
+  async #hasActiveUnreviewedSubmissions(challengeId) {
+    console.log(`Checking if there are active unreviewed submissions for challenge ${challengeId}`);
+    return Promise.resolve(false);
+  }
+
+  async #executeRule(rule, facts) {
+    const ruleEngine = new Engine();
+    ruleEngine.addRule(rule);
+
+    const result = await ruleEngine.run(facts);
+
+    const failureReasons = result.failureResults.map((failureResult) => ({
+      rule: rule.name,
+      failedConditions: failureResult.conditions.all
+        .filter((condition) => !condition.result)
+        .map((condition) => ({
+          fact: condition.fact,
+          operator: condition.operator,
+          value: condition.value,
+        })),
+    }));
+
+    return {
+      success: result.events.length > 0,
+      failureReasons,
+    };
+  }
+
+  // prettier-ignore
+  #updateSubsequentPhases(phases, currentPhase, delta) {
```

Comment:
If this is F2F, after submission phase is closed we shouldn't update iterative phases.